builtin/logical/ssh/path_roles.go

```diff
@@ -0,0 +1,149 @@
+)
+
+func pathRoles(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "roles/(?P<name>\\w+)",
+		Fields: map[string]*framework.FieldSchema{
+			"name": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Name of the role",
+			},
+			"key": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Named key in Vault",
+			},
+			"admin_user": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Admin user at target address",
+			},
+			"default_user": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Default user to whom the dynamic key is installed",
+			},
+			"cidr": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "CIDR blocks and IP addresses",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.WriteOperation:  b.pathRoleWrite,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) pathRoleWrite(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+	keyName := d.Get("key").(string)
+	adminUser := d.Get("admin_user").(string)
+	defaultUser := d.Get("default_user").(string)
+	cidr := d.Get("cidr").(string)
+
+	//input validations
+	if roleName == "" {
+		return logical.ErrorResponse("Invalid 'roleName'"), nil
+	}
+	if keyName == "" {
+		return logical.ErrorResponse("Invalid 'key'"), nil
+	}
+	if adminUser == "" {
+		return logical.ErrorResponse("Invalid 'admin_user'"), nil
+	}
+	if cidr == "" {
+		return logical.ErrorResponse("Invalid 'cidr'"), nil
+	}
+	for _, item := range strings.Split(cidr, ",") {
+		_, _, err := net.ParseCIDR(item)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Invalid cidr entry '%s'", item)), nil
+		}
+	}
+
+	keyPath := "keys/" + keyName
+	keyEntry, err := req.Storage.Get(keyPath)
+	if err != nil || keyEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Invalid 'key': '%s'", keyName)), nil
+	}
+
+	if defaultUser == "" {
+		defaultUser = adminUser
+	}
+
+	rolePath := "policy/" + roleName
+	entry, err := logical.StorageEntryJSON(rolePath, sshRole{
+		KeyName:     keyName,
+		AdminUser:   adminUser,
+		DefaultUser: defaultUser,
+		CIDR:        cidr,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+	rolePath := "policy/" + roleName
+	entry, err := req.Storage.Get(rolePath)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"policy": string(entry.Value),
```

Comment:
It shouldn't just dump it in like this. You want the specific keys like `default_user` and `admin_user` `cidr` etc